baselines/fedvssl/fedvssl/main.py

```diff
@@ -0,0 +1,268 @@
+from hydra.utils import instantiate
+from mmengine.config import Config
+from omegaconf import DictConfig, OmegaConf
+
+from .client import SslClient
+from .CtP.pyvrl.builder import build_model
+from .utils import init_p_paths, load_data, load_model, set_config_mmcv
+
+
+def t_f(arg):
+    """Transfer bool format."""
+    label = str(arg).upper()
+    if "TRUE".startswith(label):
+        return True
+    elif "FALSE".startswith(label):
+        return False
+
+
+def initial_setup(
+    cid, base_work_dir, rounds, data_dir, num_gpus, partition_dir, cfg_path
+):
+    """Initialise setup for instantiating client class."""
+    cid_plus_one = str(int(cid) + 1)
+    args = Namespace(
+        cfg=cfg_path,  # Path to the pretraining configuration file
+        checkpoint=None,
+        cid=int(cid),
+        data_dir=data_dir,
+        gpus=num_gpus,
+        launcher="none",
+        local_rank=0,
+        progress=False,
+        resume_from=None,
+        rounds=rounds,
+        seed=7,
+        validate=False,
+        work_dir=base_work_dir + "/client" + cid_plus_one,
+    )
+
+    print("Starting client", args.cid)
+    # fetch the configuration file
+    cfg = Config.fromfile(args.cfg)
+    # define the client data files;
+    # usually contains paths to the samples and annotations
+    cfg.data.train.data_source.ann_file = (
+        partition_dir + "/client_dist" + cid_plus_one + ".json"
+    )
+
+    distributed, logger = set_config_mmcv(args, cfg)
+    # These two arguments needs to be false during federated pretraining
+    # otherwise the client will load the previously saved checkpoint
+
+    cfg.resume_from = False
+    cfg.load_from = False
+
+    # load the model
+    model = load_model(cfg)
+    # load the training data
+    train_dataset = load_data(cfg)
+
+    # since during pretraining of FedVSSL
+    # we don't need any testing data, we can left it empty.
+    test_dataset = " "
+
+    return args, cfg, distributed, logger, model, train_dataset, test_dataset
+
+
+def fit_config(rnd: int) -> Dict[str, str]:
+    """Return a configuration with global epochs."""
+    config = {
+        "epoch_global": str(rnd),
+    }
+    return config
+
+
+@hydra.main(config_path="conf", config_name="base", version_base=None)
+def main(cfg: DictConfig) -> None:
+    """Define the main function for FL pre-training or fine-tuning."""
+    # print config structured as YAML
+    print(OmegaConf.to_yaml(cfg))
+
+    if cfg.pre_training:
+        # first the paths needs to be defined
+        # otherwise the program may not be able to locate the files of the ctp
+
+        init_p_paths("fedvssl")
+
+        client_resources = {
+            "num_cpus": cfg.cpus_per_client,
+            "num_gpus": cfg.gpus_per_client,
+        }
+        base_work_dir = cfg.exp_name
+        rounds = cfg.rounds
+        data_dir = cfg.data_dir
+        partition_dir = cfg.partition_dir
+        num_gpus = cfg.gpus_per_client
```

Comment:
this will create issues if `gpus_per_client` is set to less than 1.0 (e.g. if we want to pack 2 clients on the same gpu for instance with `gpus_per_client: 0.5`.

Reply:
@yan-gao-GY   

`num_gpus = int(np.ceil(cfg.client_resources.num_gpus))` # will resolve the issue.